### PR TITLE
User Profile: Add new system index for profile documents

### PR DESCRIPTION
**Pull request by ywangd**
This PR adds a new system index, '.security-profile' to store the user
profile documents. It is managed by the security plugin along side with
other security related indices. It has similar index settings (most
importantly 1 shard and 0-1 replication) and a basic mapping (based on
previous discussion) to enable further developements.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 1]
```diff
@@ -38,25 +38,33 @@
 
     public static final int INTERNAL_MAIN_INDEX_FORMAT = 6;
     private static final int INTERNAL_TOKENS_INDEX_FORMAT = 7;
+    private static final int INTERNAL_PROFILE_INDEX_FORMAT = 8;
+
+    public static final String INTERNAL_SECURITY_PROFILE_INDEX_8 = ".security-profile-8";
+    public static final String SECURITY_PROFILE_ALIAS = ".security-profile";
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:44]: I chose to not place these names alongside with other security index names in `RestrictedIndicesNames` class because `RestrictedIndicesNames` is likely obsolete since we have made all system indices also restricted. There are tests need to be fixed before we can remove the class. In the meantime, I don't think we need add anything more into it.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 2]
```diff
@@ -683,4 +696,139 @@ private XContentBuilder getTokenIndexMappings() {
...
+                builder.field("dynamic", "strict");
+                builder.startObject("properties");
+                {
+                    builder.startObject("uid");
+                    builder.field("type", "keyword");
+                    builder.endObject();
+
+                    builder.startObject("redirect_uid");
+                    builder.field("type", "keyword");
+                    builder.endObject();
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:743]: This is a new field that is not in the tech spec. I think it could be useful for migration purpose, e.g. a separate profile was somehow created for the same person and already used in case assignment etc. A lightweight migration can be setting this field to point to the old profile document id. Happy to drop this field if it feels like pre-mature optimisation.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:743]: I don't have enough information to tell how I feel about it. If this has been discussed before, please let me know.
Otherwise, we can discuss now, or at any other time you require my input.

I undestand that this is used to "link" two different profiles across different security domains. I would also like to understand when/how such profiles can be linked, un-linked, and how listing and searching works for linked profiles.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 3]
```diff
@@ -683,4 +696,139 @@ private XContentBuilder getTokenIndexMappings() {
...
+
+                    builder.startObject("enabled");
+                    builder.field("type", "boolean");
+                    builder.endObject();
+
+                    builder.startObject("user");
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:749]: Assuming this is up for discussion, I would think `profile` is more suitable.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:743]: I am happy to drop the field since it does not worth to stall the PR over it. We can always add it later if at all.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 3 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:749]: My take is that `profile` means the whole document and this section is the `user` of the profile. So I prefer keeping it as is.

**Comment by ywangd** [comment]
@albertzaharovits I removed the `redirect_uid` field. On hindsight, it was premature and unnecessary. It may be useful but we should spend some time to discuss about it and that discussion can happen later when basic functions of this feature are in place.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 4]
```diff
@@ -683,4 +696,139 @@ private XContentBuilder getTokenIndexMappings() {
...
+                        builder.endObject();
+                    }
+                    builder.endObject();
+
+                    builder.startObject("last_synchronized");
+                    builder.field("type", "date");
+                    builder.field("format", "epoch_millis");
+                    builder.endObject();
+
+                    builder.startObject("access");
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:796]: What do you think about `last_login` instead of `access`? Just a suggestion.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 5]
```diff
@@ -683,4 +696,139 @@ private XContentBuilder getTokenIndexMappings() {
...
+                                    builder.startObject("domain");
+                                    builder.field("type", "keyword");
+                                    builder.endObject();
+                                }
+                                builder.endObject();
+                            }
+                            builder.endObject();
+
+                            builder.startObject("email");
+                            builder.field("type", "text");
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:776]: There are examples in the main security index where we use the [UAX URL email tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-uaxurlemail-tokenizer.html) for this.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 6]
```diff
@@ -683,4 +696,139 @@ private XContentBuilder getTokenIndexMappings() {
...
+                            builder.endObject();
+
+                            builder.startObject("full_name");
+                            builder.field("type", "text");
+                            builder.endObject();
+
+                            builder.startObject("display_name");
+                            builder.field("type", "text");
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:784]: I think we can take advantage of the [Search-as-you-type field type](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-as-you-type.html) for all the fields that should be searchable in the textbox (`username`, `full_name` and `display_name` from my understanding).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 7]
```diff
@@ -683,4 +696,139 @@ private XContentBuilder getTokenIndexMappings() {
...
+                    builder.endObject();
+
+                    builder.startObject("last_synchronized");
+                    builder.field("type", "date");
+                    builder.field("format", "epoch_millis");
+                    builder.endObject();
+
+                    builder.startObject("access");
+                    {
+                        builder.field("type", "object");
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:798]: image

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 1 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:44]: SGTM.

**Review by albertzaharovits — COMMENTED** [review]
I had a couple of suggestions around the mappings.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 5 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:776]: Thanks for the pointer. I updated it to be the same as the email field of the main security index. I also wonder whether this could just be a keyword field because the tokenizer is more useful when the input is a text block. But what we have here for this field is likely just the email itself which does not require further analysis. That said, I like it being consistent with the settings used in the main security index.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 6 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:784]: That's a good suggestion, thanks! It shows how much (little) I know about our search offerings. I was going to leave `username` as `keyword` since it is really meant to be a single word. But on a second thought, we do allow spaces to be used in usernames and usernames from external realms have even greater flexibility. In summary, I changed all three fields to be `search_as_you_type`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 7 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:798]: Since we decided to not let past logins to affect user roles and privileges (search related), I think we should treat these past login event data as any other application specific data. Concretely, this means they are stored under the `application_data` field. Applications (Kibana in this case) can retrieve them and give them the meanings when presenting to users. But from ES's perspective, they are just another set of data. Since we don't provide search for them, we don't need to handle them specially.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 4 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:796]: I pondered on this and my conclusion is to stick with `access`. A few reasons:
* The whole concept of profile in ES has the assumption of last login because currently it is not possible to know anything about users from some external realms. So this assumption is implicit but known.
* Strictly speaking, the fields `email`, `full_name` are also from last login. But I feel it is too verbose to qualify all of them. We may also wanna make special case for `full_name` to not take value from the last login if it is `null`, i.e. we don't overwrite an existing value with `null`. In that case, it's hard to give this field a proper qualifier. 
* We talked about the possibilities to update these fields without the need of login. This is already technically possible today for native realm users. It might be possible in future to do it for more realm types. In these situations, the name `last_login` would become obsolete and inaccurate.

I do agree it is important to call out that these fields mostly reflect the values from last login. But I think we can solve it with documentation and other communications, especially given the consumers of this document are mostly internal.

**Comment by ywangd** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 3 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:750]: Okay, makes sense, thanks.
Do  you think nesting all these under a `profile` object would be better? From past experience I learned that we might end up storing multiple object types in the same index.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 5 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:781]: >  also wonder whether this could just be a keyword

I would lean to the `search-as-you-type` field for this too, because sometimes the email works as a name.
We don't have to settle it know, we should experiment with building the query for a multitude of such fields.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 7 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:804]: SGTM.

**Review by albertzaharovits — APPROVED** [review]
We agreed that the mapping is not something we should set in stone at this moment.
I'm approving knowing that we'll have to iterate over it later when we shape the APIs.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 3 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:750]: I'd say No mainly because I think we are moving away (or at least want to move away) from storing multiple different types of documents in a single index. I am aware that the main security index is an example of hosting different type of documents. But I don't think it is something we should model after today. In fact, we had previous discussions about moving API keys out of the main security index (https://github.com/elastic/elasticsearch/issues/69685). The current general indexing strategy also favours smaller and more concrete indices. I also didn't add a `doc_type` field to the mapping along this line of thinking.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java (hunk 5 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:781]: I am not sure whether `search_as_you_type` is helpful here. Based on my understanding, it helps when the value has multiple tokens and you'd like to search these tokens in any order with prefix match. For an email, we can either treat it  as a single token or multiple tokens. In the case of single token, `search_as_you_type` does not really offer anything because there is only a single token (it might still be better at performance, not sure). 

In the case of multiple tokens, it may not make a lot sense because email `foo@example.com` maybe analyzed into `foo`, `example`, `com` and matching the later two tokens is unlikely useful.

In summary, I'll keep it as is for now and we can always come back to it later.

**Comment by ywangd** [comment]
> We agreed that the mapping is not something we should set in stone at this moment. I'm approving knowing that we'll have to iterate over it later when we shape the APIs.

Thanks Albert. Yes we can and most likely will iterate on the mapping along the way.

**Comment by ywangd** [comment]
I am merging this PR because Albert approved and Tim is also happy for it to go through unless there is any contentious topics. Again, the mapping is not final. But we can iterate that through the process.
